gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/matching/single/cypher/planning/planner/greedy/GreedyPlanner.java

```diff
@@ -0,0 +1,394 @@
+   */
+  private final GraphStatistics graphStatistics;
+  /**
+   * The morphism type for vertex mappings.
+   */
+  private final MatchStrategy vertexStrategy;
+  /**
+   * The morphism type for edge mappings.
+   */
+  private final MatchStrategy edgeStrategy;
+
+  /**
+   * Creates a new greedy planner.
+   *
+   * @param graph search graph
+   * @param queryHandler query handler
+   * @param graphStatistics search graph statistics
+   * @param vertexStrategy morphism type for vertex mappings
+   * @param edgeStrategy morphism type for edge mappings
+   */
+  public GreedyPlanner(LogicalGraph graph, QueryHandler queryHandler,
+    GraphStatistics graphStatistics, MatchStrategy vertexStrategy, MatchStrategy edgeStrategy) {
+    this.graph = graph;
+    this.queryHandler = queryHandler;
+    this.graphStatistics = graphStatistics;
+    this.vertexStrategy = vertexStrategy;
+    this.edgeStrategy = edgeStrategy;
+  }
+
+  /**
+   * Computes the {@link PlanTableEntry} that wraps the {@link QueryPlan} with the minimum costs
+   * according to the greedy optimization algorithm.
+   *
+   * @return entry with minimum execution costs
+   */
+  public PlanTableEntry plan() {
+    PlanTable planTable = initPlanTable();
+
+    while (planTable.size() > 1) {
+      PlanTable newPlans = evaluateJoins(planTable);
+      newPlans = evaluateFilter(newPlans);
+      newPlans = evaluateProjection(newPlans);
+      // get plan with minimum costs and remove all plans covered by this plan
+      PlanTableEntry bestEntry = newPlans.min();
+      planTable.removeCoveredBy(bestEntry);
+      planTable.add(bestEntry);
+    }
+
+    return planTable.get(0);
+  }
+
+  //------------------------------------------------------------------------------------------------
+  // Initialization
+  //------------------------------------------------------------------------------------------------
+
+  /**
+   * Creates the initial plan table entries according to the specified vertices and edges.
+   *
+   * @return initial plan table
+   */
+  private PlanTable initPlanTable() {
+    PlanTable planTable = new PlanTable();
+    planTable = createVertexPlans(planTable);
+    planTable = createEdgePlans(planTable);
+    return planTable;
+  }
+
+  //------------------------------------------------------------------------------------------------
+  // Leaf nodes (i.e. vertices and (variable length) edges)
+  //------------------------------------------------------------------------------------------------
+
+  /**
+   * Creates an initial {@link PlanTableEntry} for each vertex in the query graph and adds it to the
+   * specified {@link PlanTable}. The entry wraps a query plan that filters vertices based on their
+   * predicates and projects properties that are required for further query planning.
+   *
+   * @param planTable plan table
+   * @return updated plan table with entries for vertices
+   */
+  private PlanTable createVertexPlans(PlanTable planTable) {
```

Comment:
Returning PlanTable imho gives the impression that a new plan tables is generated, while the actual update happens in situ